fix: refund unused ETH from EOAs after tests end

avorylli opened the pull request:
<!--
Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md
-->

**Description**

Implements automatic refund of unused ETH from EOAs back to address 0x0 after tests complete.

It prevents the faucet from being depleted during long-running tests. The implementation:
- Tracks funded addresses in a global registry
- Registers cleanup handlers that transfer remaining funds
- Leaves small reserve (0.01 ETH) for gas fees

**Tests**

No explicit tests added, but code passes all compilation and static analysis checks. Functionality works by registering Go's t.Cleanup() handlers, which automatically execute after tests complete.

**Additional context**

Refund functionality is implemented in EOA rather than Faucet, since EOA has access to the necessary transaction methods and knows its own balanc

**Metadata**

Fixes #15930

avorylli committed:
Update eoa.go

avorylli committed:
Update faucet.go

protolambda reviewed (changes requested):
Sorry to block this, but I believe this is not the right approach to test-ETH refunding.

We should not make refunds part of test execution. The mnemonic is known, the keys are known. We can access the keys after execution.

If the environment is persistent we can just sweep the funds from the known keys, using some tool or automation after test execution, to collect the test ETH dust.

If we do this as `t.Cleanup` in tests, then we only make test execution take extra long, and then we still can't be sure that the test ETH gets collected.

Besides, in some backends the environment is entirely ephemeral (e.g. sysgo), and cleaning up of ETH is not needed, as the whole chain won't exist anymore after execution completes.
Similarly, in Kurtosis we also don't really have to clean up, since Kurtosis is also not that persistent. Kurtosis ETH is very cheap.

It's only really in the real-world network where it matters.
@sigma was working on faucet key derivation and has some opinions too. 

We ideally just re-generate the keys, and sweep them all, using some command that can be directed at the persistent network that the tests ran against.

opgitgovernance commented:
This pr has been automatically marked as stale and will be closed in 5 days if no updates

avorylli committed:
Update eoa.go

avorylli committed:
Update faucet.go

avorylli committed:
Create main.go

avorylli commented:
@protolambda Thanks for the detailed review, really appreciate it!

I've reworked this completely. Instead of doing cleanup during tests with `t.Cleanup()`, I created a standalone CLI utility that can sweep funds after tests finish.

**What's different now:**
- Removed all the cleanup logic from EOA and faucet code
- Built `op-devstack/cmd/sweep-funds` tool that uses the known mnemonic to derive keys
- Tool runs separately after tests, so no performance hit during test execution
- Only meant for persistent environments where ETH actually matters

**Basic usage:**
```bash
go run ./op-devstack/cmd/sweep-funds \
  -rpc="http://localhost:8545" \
  -mnemonic="test mnemonic here" \
  -chain-id=901
```

Let me know if this direction looks better!
